nist_fanuc/doc/doxygen/man/man3/Crcl.3
.TH "Crcl" 3 "Thu Apr 14 2016" "CRCL FANUC" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Crcl \- 
.SH SYNOPSIS
.br
.PP
.SS "Classes"

.in +1c
.ti -1c
.RI "struct \fBGripperStatus\fP"
.br
.RI "\fI\fBGripperStatus\fP dummy class for future gripper information\&. \fP"
.ti -1c
.RI "struct \fBJointReport\fP"
.br
.RI "\fI\fBJointReport\fP dummy class for future customization of \fBCrcl\fP status reports\&. \fP"
.ti -1c
.RI "struct \fBCrclStatus\fP"
.br
.RI "\fI\fBCrclStatus\fP is a class that encapsulates all the CRCL information\&. All lot of the knowledge is converting ROS oriented \fBRCS\fP data into codesynthesis \fBCrcl\fP representation and vice versa\&. \fBCrclStatus\fP maintains the unit a crcl session uses to transmit robot commands, latest robot status, etc\&. \fP"
.ti -1c
.RI "class \fBCrclDelegateInterface\fP"
.br
.RI "\fI\fBCrclDelegateInterface\fP parses \fBCrcl\fP commands and interprets into robot motion\&. https://github.com/usnistgov/crcl/blob/master/doc/Reference.md\&. \fP"
.ti -1c
.RI "class \fBCrclClientCmdInterface\fP"
.br
.RI "\fI\fBCrclClientCmdInterface\fP generates \fBCrcl\fP XML command message from to send to a \fBCrcl\fP server\&. \fP"
.ti -1c
.RI "class \fBCrclStatusMsgInterface\fP"
.br
.RI "\fI\fBCrclStatusMsgInterface\fP parses a \fBCrcl\fP XML status message from a \fBCrcl\fP server\&. \fP"
.in -1c
.SS "Typedefs"

.in +1c
.ti -1c
.RI "typedef \fBRCS::Vector3\fP \fBVector3D\fP"
.br
.in -1c
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fBCRCLCmdStatus\fP { \fBCRCL_Done\fP = 0, \fBCRCL_Error\fP, \fBCRCL_Working\fP, \fBCRCL_Ready\fP }"
.br
.ti -1c
.RI "enum \fBCrclReturn\fP { \fBCANON_REJECT\fP = -2, \fBCANON_FAILURE\fP = -1, \fBCANON_SUCCESS\fP = 0, \fBCANON_STATUSREPLY\fP = 1, \fBCANON_MOTION\fP = 2, \fBCANON_RUNNING\fP, \fBCANON_NOT_IMPLEMENTED\fP }"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "bool \fBGetRPY\fP (\fBRCS::Vector3\fP Xrot, \fBRCS::Vector3\fP Zrot, double &dRoll, double &dPitch, double &dYaw)"
.br
.ti -1c
.RI "\fBPoseType\fP \fBNullPose\fP ()"
.br
.ti -1c
.RI "\fBCrcl::VectorType\fP \fBVectorZero\fP ()"
.br
.ti -1c
.RI "\fBCrcl::PoseType\fP \fBIdentityPose\fP ()"
.br
.ti -1c
.RI "\fBRCS::Vector3\fP \fBGetVector3D\fP (\fBCrcl::VectorType\fP &vector)"
.br
.ti -1c
.RI "\fBRCS::Vector3\fP \fBGetVector3D\fP (\fBCrcl::PointType\fP &point)"
.br
.ti -1c
.RI "tf::Vector3 \fBGetTfVector\fP (\fBCrcl::PointType\fP &point)"
.br
.ti -1c
.RI "tf::Vector3 \fBGetTfVector\fP (\fBCrcl::VectorType\fP &vector)"
.br
.ti -1c
.RI "inline::PointType \fBGetPoint\fP (\fBRCS::Vector3\fP &point)"
.br
.ti -1c
.RI "inline::VectorType \fBGetVector\fP (\fBRCS::Vector3\fP &point)"
.br
.ti -1c
.RI "tf::Matrix3x3 \fBGetTfRotMatrix\fP (\fBRCS::Vector3\fP Xrot, \fBRCS::Vector3\fP Zrot)"
.br
.ti -1c
.RI "\fBRCS::Pose\fP \fBConvert\fP (\fBCrcl::PoseType\fP &pose, double lengthConversion=1\&.0)"
.br
.RI "\fIConvert converts codesynthesis \fBCrcl\fP Pose into ROS/RCS pose using length conversion \&. Note, no angle conversion - rotation/orientation always in radians\&. \fP"
.ti -1c
.RI "bool \fBGetPoseToRPY\fP (\fBCrcl::PoseType\fP &pose, double &dRoll, double &dPitch, double &dYaw)"
.br
.RI "\fIGet roll,pitch,yaw orientation from codesynthesis \fBCrcl\fP pose\&. \fP"
.ti -1c
.RI "\fBRCS::Rotation\fP \fBConvert\fP (\fBRCS::Vector3\fP Xrot, \fBRCS::Vector3\fP Zrot)"
.br
.RI "\fIConvert converts codesynthesis \fBCrcl\fP rotation (as x,z rotation vectors) into ROS/RCS rotation\&. \fP"
.ti -1c
.RI "\fBCrcl::PoseType\fP \fBConvert\fP (\fBRCS::Pose\fP pose)"
.br
.RI "\fIConvert converts ROS/RCS pose into codesynthesis \fBCrcl\fP Pose\&. Note, no angle conversion - rotation/orientation always in radians\&. \fP"
.ti -1c
.RI "std::string \fBDumpCrclJoints\fP (\fBCrcl::JointStatusSequence\fP jin)"
.br
.ti -1c
.RI "\fBsensor_msgs::JointState\fP \fBConvert\fP (\fBCrcl::JointStatusSequence\fP jointStatusSeq, double angleConversion=1\&.0)"
.br
.RI "\fIConvert converts codesynthesis status actuator joint sequence into ROS JointState\&.(Fills position)\&. \fP"
.ti -1c
.RI "\fBCrcl::JointStatusSequence\fP \fBConvert\fP (\fBJointState\fP joints, double _angleConversion)"
.br
.ti -1c
.RI "\fBJointStatusSequence\fP \fBConvert\fP (\fBCrcl::ActuatorJointSequence\fP joints, double _angleConversion)"
.br
.ti -1c
.RI "\fBCrcl::PoseType\fP \fBPoseHome\fP ()"
.br
.RI "\fICreate codesynthesis \fBCrcl\fP pose that is necessary for all codesynthesis \fBCrcl\fP pose constructors\&. \fP"
.ti -1c
.RI "std::vector< double > \fBConvertToAnglePositionVector\fP (\fBCrcl::ActuatorJointSequence\fP &joints, double dAngleConversion)"
.br
.ti -1c
.RI "std::string \fBDumpRotationAsCrcl\fP (\fBRCS::Pose\fP rcspose, std::string separator)"
.br
.ti -1c
.RI "std::string \fBDumpRotationAsCrcl\fP (\fBCrcl::PoseType\fP pose, std::string separator)"
.br
.ti -1c
.RI "std::string \fBDumpCrclCommand\fP (::CRCLCommandType &crclCommand)"
.br
.ti -1c
.RI "std::string \fBDumpStatusReply\fP (\fBCrclStatus\fP *wm)"
.br
.ti -1c
.RI "std::string \fBDumpPosition\fP (\fBCrcl::PoseType\fP pose, std::string separator=',')"
.br
.RI "\fIDump contents of codesynthesis \fBCrcl\fP pose\&. \fP"
.ti -1c
.RI "std::string \fBDumpPose\fP (\fBCrcl::PoseType\fP pose, std::string separator)"
.br
.ti -1c
.RI "std::vector< double > \fBConvertToPositionVector\fP (\fBActuatorJointSequence\fP &, double dConversion)"
.br
.RI "\fIConvertToPositionVector converts codesynthesis actuator sequence into std vector of position as doubles\&. \fP"
.ti -1c
.RI "\fBJointStatusSequence\fP \fBConvert\fP (\fBActuatorJointSequence\fP jin)"
.br
.RI "\fIConvert converts codesynthesis commanded actuator joint sequence into status actuator joint sequence \&. \fP"
.ti -1c
.RI "\fBCrcl::JointStatusSequence\fP \fBConvert\fP (\fBJointState\fP joints)"
.br
.RI "\fIConvert converts ROS JointState (primarily position for now) and creates codesynthesis status actuator joint sequence\&. \fP"
.in -1c
.SS "Variables"

.in +1c
.ti -1c
.RI "typedef::ActuateJointsType::ActuateJoint_sequence \fBActuatorJointSequence\fP"
.br
.ti -1c
.RI "typedef::PoseType \fBPoseType\fP"
.br
.ti -1c
.RI "typedef::JointStatusType \fBJointStatus\fP"
.br
.ti -1c
.RI "typedef::CommandStateEnumType \fBCommandStateEnum\fP"
.br
.ti -1c
.RI "typedef::PointType \fBPointType\fP"
.br
.ti -1c
.RI "typedef::VectorType \fBVectorType\fP"
.br
.ti -1c
.RI "typedef::JointStatusesType::JointStatus_sequence \fBJointStatusSequence\fP"
.br
.ti -1c
.RI "typedef::PoseToleranceType \fBPoseToleranceType\fP"
.br
.in -1c
.SH "Typedef Documentation"
.PP 
.SS "typedef \fBRCS::Vector3\fP \fBCrcl::Vector3D\fP"

.SH "Enumeration Type Documentation"
.PP 
.SS "enum \fBCrcl::CRCLCmdStatus\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fICRCL_Done \fP\fP
.TP
\fB\fICRCL_Error \fP\fP
.TP
\fB\fICRCL_Working \fP\fP
.TP
\fB\fICRCL_Ready \fP\fP
.SS "enum \fBCrcl::CrclReturn\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fICANON_REJECT \fP\fP
.TP
\fB\fICANON_FAILURE \fP\fP
.TP
\fB\fICANON_SUCCESS \fP\fP
.TP
\fB\fICANON_STATUSREPLY \fP\fP
.TP
\fB\fICANON_MOTION \fP\fP
.TP
\fB\fICANON_RUNNING \fP\fP
.TP
\fB\fICANON_NOT_IMPLEMENTED \fP\fP
.SH "Function Documentation"
.PP 
.SS "\fBRCS::Pose\fP Crcl::Convert (\fBCrcl::PoseType\fP &pose, doublelengthConversion = \fC1\&.0\fP)"

.PP
Convert converts codesynthesis \fBCrcl\fP Pose into ROS/RCS pose using length conversion \&. Note, no angle conversion - rotation/orientation always in radians\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcodesynthesis\fP CRCL pose type\&. 
.br
\fIlengthConversion\fP to make length in meters\&. (e\&.g\&., if mm input, conversion = 0\&.001)\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
ROS/RCS pose\&. 
.RE
.PP

.SS "\fBJointStatusSequence\fP Crcl::Convert (ActuatorJointSequencejin)"

.PP
Convert converts codesynthesis commanded actuator joint sequence into status actuator joint sequence \&. 
.PP
\fBParameters:\fP
.RS 4
\fIcodesynthesis\fP sequence of commanded joints\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
codesynthesis sequence of status joints\&. 
.RE
.PP

.SS "\fBRCS::Rotation\fP Crcl::Convert (\fBRCS::Vector3\fPXrot, \fBRCS::Vector3\fPZrot)"

.PP
Convert converts codesynthesis \fBCrcl\fP rotation (as x,z rotation vectors) into ROS/RCS rotation\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcodesynthesis\fP CRCL x,z rotation vectors as used in pose\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
ROS/RCS rotation\&. 
.RE
.PP

.SS "\fBCrcl::PoseType\fP Crcl::Convert (\fBRCS::Pose\fPpose)"

.PP
Convert converts ROS/RCS pose into codesynthesis \fBCrcl\fP Pose\&. Note, no angle conversion - rotation/orientation always in radians\&. 
.PP
\fBParameters:\fP
.RS 4
\fIROS/RCS\fP pose\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
codesynthesis CRCL pose type\&. 
.RE
.PP

.SS "\fBCrcl::JointStatusSequence\fP Crcl::Convert (\fBJointState\fPjoints)"

.PP
Convert converts ROS JointState (primarily position for now) and creates codesynthesis status actuator joint sequence\&. 
.PP
\fBParameters:\fP
.RS 4
\fIROS\fP JointState with Joint positions filled\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
codesynthesis sequence of status joints\&. 
.RE
.PP

.SS "\fBsensor_msgs::JointState\fP Crcl::Convert (\fBCrcl::JointStatusSequence\fPjointStatusSeq, doubleangleConversion = \fC1\&.0\fP)"

.PP
Convert converts codesynthesis status actuator joint sequence into ROS JointState\&.(Fills position)\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcodesynthesis\fP sequence of status joints\&. 
.br
\fIconversion\fP factor for each joint (e\&.g\&., degree to radian)\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
ROS JointState with Joint positions filled\&. 
.RE
.PP

.SS "\fBCrcl::JointStatusSequence\fP Crcl::Convert (\fBJointState\fPjoints, double_angleConversion)"

.SS "\fBJointStatusSequence\fP Crcl::Convert (\fBCrcl::ActuatorJointSequence\fPjoints, double_angleConversion)"

.SS "std::vector<double> Crcl::ConvertToAnglePositionVector (\fBCrcl::ActuatorJointSequence\fP &joints, doubledAngleConversion)"

.SS "std::vector<double> Crcl::ConvertToPositionVector (ActuatorJointSequence &, doubledConversion)"

.PP
ConvertToPositionVector converts codesynthesis actuator sequence into std vector of position as doubles\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcodesynthesis\fP sequence of joints\&. 
.br
\fIconversion\fP factor for each joint (e\&.g\&., degree to radian)\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
std vector of doubles representing position\&. 
.RE
.PP

.SS "std::string Crcl::DumpCrclCommand (::CRCLCommandType &crclCommand)"

.SS "std::string Crcl::DumpCrclJoints (\fBCrcl::JointStatusSequence\fPjin)"

.SS "std::string Crcl::DumpPose (\fBCrcl::PoseType\fPpose, std::stringseparator)"

.SS "std::string Crcl::DumpPosition (\fBCrcl::PoseType\fPpose, std::stringseparator = \fC','\fP)"

.PP
Dump contents of codesynthesis \fBCrcl\fP pose\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcodesynthesis\fP CRCL pose\&. 
.br
\fIseparator\fP defines character to use a separater between values (e\&.g\&., ',' for csv)\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
string with \fBCrcl\fP pose contents\&. 
.RE
.PP

.SS "std::string Crcl::DumpRotationAsCrcl (\fBRCS::Pose\fPrcspose, std::stringseparator)"

.SS "std::string Crcl::DumpRotationAsCrcl (\fBCrcl::PoseType\fPpose, std::stringseparator)"

.SS "std::string Crcl::DumpStatusReply (CrclStatus *wm)"

.SS "inline ::\fBPointType\fP Crcl::GetPoint (\fBRCS::Vector3\fP &point)"

.SS "bool Crcl::GetPoseToRPY (\fBCrcl::PoseType\fP &pose, double &dRoll, double &dPitch, double &dYaw)"

.PP
Get roll,pitch,yaw orientation from codesynthesis \fBCrcl\fP pose\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcodesynthesis\fP CRCL pose\&. 
.br
\fIroll,pitch,yaw\fP reference to doubles that will be filled with angles in radians\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
true if sucessful, false otherwise\&. 
.RE
.PP

.SS "bool Crcl::GetRPY (\fBRCS::Vector3\fPXrot, \fBRCS::Vector3\fPZrot, double &dRoll, double &dPitch, double &dYaw)"

.SS "tf::Matrix3x3 Crcl::GetTfRotMatrix (\fBRCS::Vector3\fPXrot, \fBRCS::Vector3\fPZrot)"

.SS "tf::Vector3 Crcl::GetTfVector (\fBCrcl::PointType\fP &point)\fC [inline]\fP"

.SS "tf::Vector3 Crcl::GetTfVector (\fBCrcl::VectorType\fP &vector)\fC [inline]\fP"

.SS "inline ::\fBVectorType\fP Crcl::GetVector (\fBRCS::Vector3\fP &point)"

.SS "\fBRCS::Vector3\fP Crcl::GetVector3D (\fBCrcl::VectorType\fP &vector)\fC [inline]\fP"

.SS "\fBRCS::Vector3\fP Crcl::GetVector3D (\fBCrcl::PointType\fP &point)\fC [inline]\fP"

.SS "\fBCrcl::PoseType\fP Crcl::IdentityPose ()"

.SS "\fBCrcl::PoseType\fP Crcl::NullPose ()"

.SS "\fBPoseType\fP Crcl::PoseHome ()"

.PP
Create codesynthesis \fBCrcl\fP pose that is necessary for all codesynthesis \fBCrcl\fP pose constructors\&. 
.PP
\fBReturns:\fP
.RS 4
empty codesynthesis \fBCrcl\fP pose\&. 
.RE
.PP

.SS "\fBCrcl::VectorType\fP Crcl::VectorZero ()\fC [inline]\fP"

.SH "Variable Documentation"
.PP 
.SS "typedef::ActuateJointsType::ActuateJoint_sequence Crcl::ActuatorJointSequence"

.SS "typedef::CommandStateEnumType Crcl::CommandStateEnum"

.SS "typedef::JointStatusType Crcl::JointStatus"

.SS "typedef::JointStatusesType::JointStatus_sequence Crcl::JointStatusSequence"

.SS "typedef::PointType Crcl::PointType"

.SS "typedef::PoseToleranceType Crcl::PoseToleranceType"

.SS "typedef::PoseType Crcl::PoseType"

.SS "typedef::VectorType Crcl::VectorType"

.SH "Author"
.PP 
Generated automatically by Doxygen for CRCL FANUC from the source code\&.


nist_fanuc/src/crcl.cpp
// crcl.cpp

/*
 * DISCLAIMER:
 * This software was produced by the National Institute of Standards
 * and Technology (NIST), an agency of the U.S. government, and by statute is
 * not subject to copyright in the United States.  Recipients of this software
 * assume all responsibility associated with its operation, modification,
 * maintenance, and subsequent redistribution.
 *
 * See NIST Administration Manual 4.09.07 b and Appendix I.
 */
#include "crcl.h"
#include "Globals.h"
#include "CrclInterface.h"
//#include "boost/array.hpp"
#define ROWMAJOR

// http://wiki.ros.org/APIs
// http://docs.ros.org/api/tf/html/c++/annotated.html
namespace Crcl {
    static const char *sStateEnums[] = {"CRCL_Done", "CRCL_Error", "CRCL_Working", "CRCL_Ready"};

    bool GetRPY(RCS::Vector3 Xrot, RCS::Vector3 Zrot, double & dRoll, double & dPitch, double & dYaw);
    PoseType NullPose();
    Crcl::VectorType VectorZero();
    Crcl::PoseType IdentityPose();

    inline RCS::Vector3 GetVector3D(Crcl::VectorType & vector) {
        return RCS::Vector3(vector.I(), vector.J(), vector.K());
    }

    inline RCS::Vector3 GetVector3D(Crcl::PointType & point) {
        return RCS::Vector3(point.X(), point.Y(), point.Z());
    }

    inline tf::Vector3 GetTfVector(Crcl::PointType & point) {
        return tf::Vector3(point.X(), point.Y(), point.Z());
    }

    inline tf::Vector3 GetTfVector(Crcl::VectorType & vector) {
        return tf::Vector3(vector.I(), vector.J(), vector.K());
    }

    inline ::PointType GetPoint(RCS::Vector3 & point) {
        return ::PointType(point.x(), point.y(), point.z());
    }

    inline ::VectorType GetVector(RCS::Vector3 & point) {
        return ::VectorType(point.x(), point.y(), point.z());
    }

    tf::Matrix3x3 GetTfRotMatrix(RCS::Vector3 Xrot, RCS::Vector3 Zrot) {

        tf::Vector3 YRot = Zrot.cross(Xrot);

        tf::Matrix3x3 m(Xrot.getX(), YRot.getX(), Zrot.getX(),
                Xrot.getY(), YRot.getY(), Zrot.getY(),
                Xrot.getZ(), YRot.getZ(), Zrot.getZ());
        return m;
    }

    bool GetRPY(RCS::Vector3 Xrot, RCS::Vector3 Zrot, double &roll, double &pitch, double &yaw) {
        tf::Matrix3x3 m = GetTfRotMatrix(Xrot, Zrot);
        m.getRPY(roll, pitch, yaw);
        return true;

    }

    RCS::Pose Convert(Crcl::PoseType & pose, double lengthConversion) {
        RCS::Pose p;
        p.setOrigin(GetTfVector(pose.Point()) * lengthConversion);
        p.setBasis(GetTfRotMatrix(GetVector3D(pose.XAxis()), GetVector3D(pose.ZAxis())));
        return p;
    }

    inline Crcl::VectorType VectorZero() {
        return Crcl::VectorType(0.0, 0.0, 0.0);
    }

    bool GetPoseToRPY(Crcl::PoseType & pose, double & dRoll, double & dPitch, double & dYaw) {
        return GetRPY(GetVector3D(pose.XAxis()), GetVector3D(pose.ZAxis()), dRoll, dPitch, dYaw);
    }

    RCS::Rotation Convert(RCS::Vector3 Xrot, RCS::Vector3 Zrot) {
        double roll, pitch, yaw;
        GetRPY(Xrot, Zrot, roll, pitch, yaw);
        RCS::Rotation rotation;
        rotation.setRPY(roll, pitch, yaw);
        return rotation;
    }

    Crcl::PoseType Convert(RCS::Pose pose) {
        Crcl::PoseType crclPose(NullPose());
        crclPose.Point() = GetPoint(pose.getOrigin());
#if 0
        crclPose.Point().X() = pose.getOrigin().x();
        crclPose.Point().Y() = pose.getOrigin().y();
        crclPose.Point().Z() = pose.getOrigin().z();
#endif
        tf::Matrix3x3 m = pose.getBasis();
        tf::Vector3 v1, v2, v3;

        v1 = m.getColumn(0);
        v2 = m.getColumn(1);
        v3 = m.getColumn(2);
        crclPose.XAxis(GetVector(v1));
        crclPose.ZAxis(GetVector(v3));

        return crclPose;
    }

    std::string DumpCrclJoints(Crcl::JointStatusSequence jin) {
        std::stringstream str;
        for (unsigned int i = 0; i < jin.size(); i++) {
            double pos = jin[i].JointPosition().present() ? jin[i].JointPosition().get() : nan("");
            str << " [" << i << "]=" << pos;
        }
        return str.str();
    }
    
    sensor_msgs::JointState Convert(Crcl::JointStatusSequence jout, double angleConversion) {
        JointState joint;
        for (unsigned int i = 0; i < jout.size(); i++) {
            //::JointStatusType  jointstatus(jout[i ].JointNumber());
            double pos = jout[i].JointPosition().present() ? jout[i].JointPosition().get(): 0.0;
            //std::cout << "Joint[" << i << "]=" << pos << std::endl;
            joint.position.push_back(pos); // *(jout[i].JointPosition()) * angleConversion);
            // FIXME: how do you determine if there are values for velocity and force/torque
            //joint.velocity.push_back(*(jout[i].JointVelocity())); 
            joint.effort.push_back(0.0);
        }
        return joint;
    }

    Crcl::JointStatusSequence Convert(JointState joints, double _angleConversion) {
        JointStatusSequence jout;

        for (unsigned int i = 0; i < joints.position.size(); i++) {
            ::JointStatusType jointstatus(i + 1);
            jointstatus.JointPosition(joints.position[i] * _angleConversion);
            jout.push_back(jointstatus);
        }
        return jout;
    }

    JointStatusSequence Convert(Crcl::ActuatorJointSequence joints, double _angleConversion) {
        JointStatusSequence jout;

        for (unsigned int i = 0; i < joints.size(); i++) {
            ::ActuateJointType & jointcmd(joints[i]);
            unsigned long long n = jointcmd.JointNumber();
            ::JointStatusType jointstatus(jointcmd.JointNumber());
            jointstatus.JointPosition(jointcmd.JointPosition() * _angleConversion);
            jout.push_back(jointstatus);
        }
        return jout;
    }

    Crcl::PoseType NullPose() {
        Crcl::PointType point(0.0, 0.0, 0.0);
        Crcl::VectorType xrot(1.0, 0.0, 0.0);
        Crcl::VectorType zrot(0.0, 0.0, 1.0);

        return Crcl::PoseType(point, xrot, zrot);
    }

    Crcl::PoseType IdentityPose() {
        Crcl::PointType point(0.0, 0.0, 0.0);
        Crcl::VectorType xrot(1.0, 0.0, 0.0);
        Crcl::VectorType zrot(0.0, 0.0, 1.0);

        return Crcl::PoseType(point, xrot, zrot);
    }

    Crcl::PoseType PoseHome() {
        static double home[6] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};

        std::vector<double> h;
        std::copy(&home[0], &home[6], std::back_inserter(h));
        RCS::Pose pose;
        // maybe hard code this
        pose.getOrigin() = RCS::Vector3(0, 0, 0);
        pose.getRotation().setRPY(0, 0, 0);
        //return Convert(IkFastKinematics::JointsToPose(h) );
        return Convert(pose);
    }

    std::vector<double> ConvertToAnglePositionVector(Crcl::ActuatorJointSequence & joints, double dAngleConversion) {
        std::vector<double> angles;

        for (unsigned int i = 0; i < joints.size(); i++) {
            ::ActuateJointType & jointcmd(joints[i]);
            angles.push_back(jointcmd.JointPosition() * dAngleConversion);
        }
        return angles;
    }
    std::string DumpRotationAsCrcl(RCS::Pose rcspose, std::string separator) {
        Crcl::PoseType pose = Convert(rcspose);
        return DumpRotationAsCrcl(pose);
    }
     std::string DumpRotationAsCrcl(Crcl::PoseType pose, std::string separator) {
       std::stringstream str;
        str << "XAxis = ";
        str << boost::format("%8.4f") % pose.XAxis().I() << separator.c_str();
        str << boost::format("%8.4f") % pose.XAxis().J() << separator.c_str();
        str << boost::format("%8.4f") % pose.XAxis().K() << separator.c_str();
        str << "\tZAxis = ";
        str << boost::format("%8.4f") % pose.ZAxis().I() << separator.c_str();
        str << boost::format("%8.4f") % pose.ZAxis().J() << separator.c_str();
        str << boost::format("%8.4f") % pose.ZAxis().K() << separator.c_str();
        return str.str();
    }

    std::string DumpCrclCommand(::CRCLCommandType & crclCommand) {
        std::stringstream str;
        try {
            if (crclCommand.Name().present())
                str << *crclCommand.Name();
        } catch (...) {
        }
        str << "CommandID=" << crclCommand.CommandID() << std::endl;
        return str.str();
    }

    std::string DumpStatusReply(CrclStatus *wm)
    {
        std::stringstream str;
        //const char * sEnums[] = {"CRCL_Done", "CRCL_Error", "CRCL_Working", "CRCL_Ready"};
        str << "GetStatusReply=" << wm->CommandID() << ":";
        str << wm->CommandStatus() << std::endl;
        str<< "Crcl Status Joints       =" << Crcl::DumpCrclJoints(wm->_CurrentJoints).c_str()<< std::endl;
        str << "Crcl Status Position    ="  << Crcl::DumpPosition(wm->_CurrentPose).c_str()<< std::endl;
        tf::Matrix3x3 m =  GetTfRotMatrix(GetVector3D(wm->_CurrentPose.XAxis()), GetVector3D(wm->_CurrentPose.ZAxis()));
        tf::Quaternion q;
        m.getRotation(q);
        str << "Crcl Status Quaternion  ="  << q.x() <<":" << q.y() <<":" << q.z() <<":" <<  q.w() <<  std::endl;
        str << "Crcl Status Orientation ="  << Crcl::DumpRotationAsCrcl(wm->_CurrentPose).c_str()<< std::endl;
        double r,p,y; GetPoseToRPY(wm->_CurrentPose,r,p,y);
        str << "Crcl Status RPY         ="  << Globals.StrFormat("%8.4f,%8.4f,%8.4f",Rad2Deg(r),Rad2Deg(p),Rad2Deg(y)).c_str()<< std::endl;
        return str.str();
    }
     std::string DumpPosition(Crcl::PoseType pose, std::string separator) {
        std::stringstream str;
        str << boost::format("%8.4f") % pose.Point().X() << separator.c_str();
        str << boost::format("%8.4f") % pose.Point().Y() << separator.c_str();
        str << boost::format("%8.4f") % pose.Point().Z() << separator.c_str();
         return str.str();
    }   
    std::string DumpPose(Crcl::PoseType pose, std::string separator) {
        std::stringstream str;
        str << boost::format("%8.4f") % pose.Point().X() << separator.c_str();
        str << boost::format("%8.4f") % pose.Point().Y() << separator.c_str();
        str << boost::format("%8.4f") % pose.Point().Z() << separator.c_str();
        str << boost::format("%8.4f") % pose.XAxis().I() << separator.c_str();
        str << boost::format("%8.4f") % pose.XAxis().J() << separator.c_str();
        str << boost::format("%8.4f") % pose.XAxis().K() << separator.c_str();
        str << boost::format("%8.4f") % pose.ZAxis().I() << separator.c_str();
        str << boost::format("%8.4f") % pose.ZAxis().J() << separator.c_str();
        str << boost::format("%8.4f") % pose.ZAxis().K() << separator.c_str();
        return str.str();
    }
    // ----------------------------------

    CrclStatus::CrclStatus() :
    _CommandStatus("CRCL_Done"),
    _GoalPose(Crcl::PoseHome()),
    _CurrentPose(Crcl::PoseHome()) {
        _GoalJoints = JointsHome();
        _CurrentJoints = JointsHome();
        CommandID() = 1;
        StatusID() = 1;
        _lengthConversion = 1.0;

        // opposite of what documentation says are defaults
        _angleUnit = RCS::DEGREE;
        _angleConversion = M_PI / 180.0;

        _lengthUnit = RCS::METER;
        _lengthConversion = 1.0;

        _forceUnit = RCS::NEWTON;
        _forceConversion = 1.0;

        _torqueUnit = RCS::NEWTONMETER;
        _torqueConversion = 1.0;

        _bCoordinatedMotion = true;

        Rates().JointVelLimit().resize(8, DEFAULT_JOINT_MAX_VEL);
        Rates().JointAccLimit().resize(8, DEFAULT_JOINT_MAX_ACCEL);
    }

    void CrclStatus::Update(JointState & joints) {
#ifdef DEBUGJOINTSTATUSCRCLUPDATE
        std::cout << RCS::DumpJoints(joints);
#endif
        JointStatusSequence jout;

        for (unsigned int i = 0; i < joints.position.size(); i++) {

            ::JointStatusType jointstatus(i + 1);

            // Convert from radians to degrees?
            double dConverter = 1.0;

            if (_angleUnit == RCS::DEGREE) {
                dConverter = (double) (180.0 / M_PI);
            }
            jointstatus.JointPosition(dConverter * joints.position[i]);
            jout.push_back(jointstatus);
        }
        _CurrentJoints = _GoalJoints = jout;
    }
    // FIXME

    Crcl::JointStatusSequence CrclStatus::JointsHome() {
        static double home[6] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};

        if (_GoalJoints.size() > 0) {
            _GoalJoints.clear();
        }

        for (int i = 0; i < 6; i++) {
            Crcl::JointStatus j(i + 1);
            j.JointPosition(home[i]);
            j.JointVelocity(0.0);
            j.JointTorqueOrForce(0.0);
            _GoalJoints.push_back(j);
        }
        return _GoalJoints;
    }

    void CrclStatus::Update(unsigned long long _CommandID) {
        StatusID() = CommandID() = _CommandID;
    }

    void CrclStatus::Update(Crcl::CommandStateEnum state) {
        CommandStatus() = state;
    }

    void CrclStatus::Update(Crcl::JointStatusSequence & joints, RCS::TrajPointType type) {
        if (type == RCS::WAYPOINT) {
            _CurrentJoints = joints;
        } else {
            _GoalJoints = joints;
        }
    }

    void CrclStatus::Update(Crcl::PoseType & pose, RCS::TrajPointType type) {
        if (type == RCS::WAYPOINT) {
            _GoalPose = pose;
            _CurrentPose = pose;
        } else {
            _CurrentPose = pose;
            _GoalPose = pose;
        }
    }

    void CrclStatus::Update(RCS::Pose & pose) {
        _CurrentPose = Convert(pose);
#if 0
        _CurrentPose.Point().X() = pose.getOrigin().x();
        _CurrentPose.Point().Y() = pose.getOrigin().y();
        _CurrentPose.Point().Z() = pose.getOrigin().z();
        boost::array<double, 12> m = _matrixFromQuat(pose.rotation);
        Crcl::VectorType xunitv(Crcl::VectorZero());
        xunitv.I(m[4 * 0 + 0]);
        xunitv.J(m[4 * 0 + 1]);
        xunitv.K(m[4 * 0 + 2]);
        _CurrentPose.XAxis(xunitv);

        Crcl::VectorType zunitv(Crcl::VectorZero());
        zunitv.I(m[4 * 2 + 0]);
        zunitv.J(m[4 * 2 + 1]);
        zunitv.K(m[4 * 2 + 2]);
        _CurrentPose.ZAxis(zunitv);
#endif
        _GoalPose = _CurrentPose;
    }
};
